server/src/main/java/org/elasticsearch/cluster/service/ClusterApplierService.java

```diff
@@ -394,12 +410,14 @@ private void runTask(String source, Function<ClusterState, ClusterState> updateF
 
         final long startTimeMillis = threadPool.relativeTimeInMillis();
         final Recorder stopWatch = new Recorder(threadPool, slowTaskThreadDumpTimeout);
+        final TimedListener timedListener = new TimedListener(clusterApplyListener, stopWatch);
         final ClusterState newClusterState;
         try {
             try (Releasable ignored = stopWatch.record("running task [" + source + ']')) {
                 newClusterState = updateFunction.apply(previousClusterState);
             }
         } catch (Exception e) {
+            timedListener.onFailure(e);
```

Comment:
If the listener throws, the following code will not run. Probably shouldn't happen in practice. But maybe still worthwhile to wrap in try-finally?

I also wonder whether we should add more details in the log message for the time spent on applying the cluster and calling the listener?

Reply:
The `TimedListener` will measure the time spent in the listener. e.g. from the tests
```
cluster state applier task [test4] took [36s] which is above the warn threshold of [30s]: [running task [test4]] took [0ms], [listener.onResponse] took [36000ms]
```
Is that what you meant by "time spent ... calling the listener"

I will look at using one of the `ActionListener....` utils to make exception handling more robust.

Reply:
I think Yang is right, we just need a `try`/`finally` here. TBH it's an error for this listener to throw, we should probably assert that too.

Reply:
Actually no real need to propagate the exception to the caller either, it just bubbles up to the unhandled exception handler which logs it and drops it. We may as well catch and log (and assert) in `TimedListener`.

Reply:
The `ClusterApplyListener` already wraps the provided listener to prevent `onFailure` exceptions from propagating, but I added similar logic to `TimedListener#onFailure` just in case that changes and we end up with a less safe listener being passed in.